presto-orc/src/main/java/com/facebook/presto/orc/reader/NullWrappingColumnReader.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.reader;
+
+import com.facebook.presto.orc.QualifyingSet;
+import com.facebook.presto.orc.stream.BooleanInputStream;
+import com.facebook.presto.orc.stream.LongInputStream;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+import static com.google.common.base.Verify.verify;
+
+abstract class NullWrappingColumnReader
+        extends ColumnReader
+        implements StreamReader
+{
+    QualifyingSet innerQualifyingSet;
+    boolean hasNulls;
+    int innerPosInRowGroup;
+    int numInnerRows;
+    int[] nullsToAdd;
+    int numNullsToAdd;
+    // Number of elements retrieved from inner reader.
+    int numInnerResults;
+
+    protected void beginScan(BooleanInputStream presentStream, LongInputStream lengthStream)
+            throws IOException
+    {
+        super.beginScan(presentStream, lengthStream);
+        numInnerResults = 0;
+    }
+
+    // Translates the positions of inputQualifyingSet between
+    // beginPosition and endPosition into an inner qualifying set for
+    // the non-null content.
+    protected void makeInnerQualifyingSets(int beginPosition, int endPosition)
+    {
+        if (presentStream == null) {
+            numInnerRows = endPosition - beginPosition;
+            hasNulls = false;
+            return;
+        }
+        hasNulls = true;
+        numInnerRows = 0;
+        if (innerQualifyingSet == null) {
+            innerQualifyingSet = new QualifyingSet();
+        }
+        QualifyingSet input = inputQualifyingSet;
+        int[] inputRows = input.getPositions();
+        int numActive = Math.min(endPosition, input.getPositionCount());
+        QualifyingSet inner = innerQualifyingSet;
+        int[] innerRows = inner.getMutablePositions(numActive);
```

Comment:
inconsistent naming: rows vs. positions; let's rename innerRows to innerPositions